### PR TITLE
Use the fetch stack for documents.

### PR DESCRIPTION
**Pull request by Ms2ger**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13742)

### CODE REVIEW

**Comment by Ms2ger** [comment]
Status update: investigating the timeout in `/_mozilla/mozilla/referrer-policy/unset-referrer-policy/meta-referrer/same-origin/http-https/iframe-tag/upgrade-protocol.keep-origin-redirect.http.html`. Apparently the load event is not dispatched at the iframe created in `queryIframe`.

**Comment by Ms2ger** [comment]
The issue lies within this lovely piece of code:
```rust
        let mut ssl_error = None;
        let metadata = match meta_result {
            Ok(meta) => Some(meta),
            Err(NetworkError::SslValidation(url, reason)) => {
                ssl_error = Some(reason);
                let mut meta = Metadata::default(url);
                let mime: Option<Mime> = "text/html".parse().ok();
                meta.set_content_type(mime.as_ref());
                Some(meta)
            },
            Err(_) => None,
        };
```

With the old stack, we get `Err(NetworkError::SslValidation(..))`, create a new `Metadata` and go on our merry way. With the fetch stack, we get some other `Err(..)` value, `ScriptThread::page_headers_available` notices `metadata` is `None` and returns `None` itself, so we bail out without a parser object, and the load event is never fired.

@jdm, thoughts?

**Comment by jdm** [comment]
I propose that we modify the `ResponseType::Error` variant to store a `NetworkError` value, and modify the block in `http_network_fetch` that gets an Err value from `obtain_response` to turn the response into an error response that contains the actual error that was returned from the network stack. This will allow the `metadata` method to return an error that was extracted from the error response, rather than just returning an internal error.

**Comment by Ms2ger** [comment]
Rebased; planning to look into your suggestion tomorrow. Thanks!

File: components/net_traits/response.rs [hunk 1]
```diff
@@ -135,11 +135,18 @@ impl Response {
 
     pub fn is_network_error(&self) -> bool {
         match self.response_type {
-            ResponseType::Error => true,
+            ResponseType::Error(..) => true,
             _ => false
         }
     }
 
+    pub fn get_network_error(&self) -> Option<&NetworkError> {
```
Comment (jdm) [inline components/net_traits/response.rs]: Every caller clones the error, so I'm not sure it's worth returning a borrow.

File: components/net/fetch/methods.rs [hunk 2]
```diff
@@ -1148,8 +1148,14 @@ fn http_network_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
     // TODO this step isn't possible yet
     // Step 13
 
+    // Step 14.
+    if credentials_flag {
+        // TODO: should block cookies?
```
Comment (jdm) [inline components/net/fetch/methods.rs]: I don't think this is a meaningful TODO given https://tools.ietf.org/html/rfc6265#section-7.

File: components/script/script_thread.rs [hunk 3]
```diff
@@ -2114,30 +2114,29 @@ impl ScriptThread {
...
+            origin: load_data.url,
             pipeline_id: Some(id),
-            credentials_flag: true,
+            referrer_url: load_data.referrer_url,
             referrer_policy: load_data.referrer_policy,
-            referrer_url: load_data.referrer_url
-        }, LoadConsumer::Listener(response_target), None)).unwrap();
+            headers: load_data.headers,
+            body: load_data.data,
+            .. RequestInit::default()
```
Comment (jdm) [inline components/script/script_thread.rs]: Let's file a followup issue to support non-follow redirect modes for RequestInit.

**Review by jdm — CHANGES_REQUESTED** [review]
Nice work; this was a straightforward series of commits to read through.

File: components/net_traits/response.rs (hunk 1 above)
Comment (Ms2ger) [inline components/net_traits/response.rs]: I think it's better to at least make it possible for new callers to not have to clone, if they don't need to.

**Pull request merged**